core/cont/inc/RangeCast.h

```diff
@@ -0,0 +1,185 @@
+// Author: Jonas Rembser, CERN, Jul 2021
+
+/*************************************************************************
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RangeCast_h
+#define ROOT_RangeCast_h
+
+#include <cassert>
+#include <type_traits>
+#include <utility>
+
+namespace ROOT {
+namespace Internal {
+
+#if (__cplusplus < 201700L)
+
+template <typename T, bool isDynamic = false>
+struct Cast {
+   template <typename U>
+   static T cast(U &&u)
+   {
+      return dynamic_cast<T>(u);
+   }
+};
+
+template <typename T>
+struct Cast<T, false> {
+   template <typename U>
+   static T cast(U &&u)
+   {
+      assert(dynamic_cast<T>(u));
+      return static_cast<T>(u);
+   }
+};
+
+#endif
+
+// For SFINAE-based checks for the existence of the `begin` and `end` methods.
+template <typename T>
+constexpr auto hasBeginEnd(int) -> decltype(std::declval<T>().begin(), std::declval<T>().end(), true)
+{
+   return true;
+}
+
+template <typename>
+constexpr bool hasBeginEnd(...)
+{
+   return false;
+}
+
+} // namespace Internal
+} // namespace ROOT
+
+template <typename T, typename WrappedIterator_t, bool isDynamic>
+class TypedIter {
```

Comment:
All subsequent declarations should still be in the `ROOT` namespace.
Probably all by the `RangeCast` functions should even be in `ROOT::Internal`, or at least `ROOT::Detail`?